### PR TITLE
fix(profiling): default to libdatadog if autoinject/SSI is enabled

### PR DESCRIPTION
**Pull request by sanchda**
In injected configurations, the Profiler can cause deployed (bundled?  stashed?) dependencies to over-ride the dependencies of the underlying application.  This is especially problematic for the `protobuf` library.

This PR changes several things
* Introduces a new "required" option to the libdatadog (libdd) collector/exporter.  Normally, if the user merely enables libdatadog, the profiler will make a best effort at using it, but fallback to the old collector/exporter if needed.  When libdatadog is required, profiling will be disabled if libdatadog cannot be used.
* Has lib-inection set libdatadog to "required"

## Checklist

- [X] Change(s) are motivated and described in the PR description
- [X] Testing strategy is described if automated tests are not included in the PR
- [X] Risks are described (performance impact, potential for breakage, maintainability)
- [X] Change is maintainable (easy to change, telemetry, documentation)
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [X] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [X] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/profiling/scheduler.py [hunk 1]
```diff
@@ -63,9 +60,9 @@ def flush(self):
         for exp in self.exporters:
             try:
                 exp.export(events, start, self._last_export)
-            except exporter.ExportError as e:
-                LOG.warning("Unable to export profile: %s. Ignoring.", _traceback.format_exception(e))
-            except Exception:
+            except Exception as e:
+                if e.__class__.__name__ == "ExportError":
```
Comment (P403n1x87) [inline ddtrace/profiling/scheduler.py:64]: Is the idea here that we want both logs on `ExportError`? Can't we do
```suggestion
                if isinstance(e, exporter.ExportError):
```
here? 🤔

Comment (sanchda) [reply ddtrace/profiling/scheduler.py]: OK, I think this is better now.

I originally removed the exporter imports out of ignorance.  I was overly pessimistic about how dependencies were specified in the module.  Instead, the top-level module just specifies its types and leaves specific loading to the caller, so it's safe to bring it back.

I verified that this patch does not erroneously load protobuf if libdd is enabled.

File: lib-injection/sitecustomize.py [hunk 2]
```diff
@@ -77,6 +77,10 @@ def _inject():
             _log("failed to load ddtrace module: %s" % e, level="error")
             return
         else:
+            # Similar to SSI, this environment variable allows downstream components to know that the library was
```
Comment (sanchda) [inline lib-injection/sitecustomize.py:80]: @Kyle-Verhoog, any bad smells from this?  I tried to align things with how SSI works at the top level, but I guess we'll clobber whatever value is here (in particular:  if the environment variables changes from a truthy-ish value to some kind of enum or whatever).

File: releasenotes/notes/profiling-fix-protobuf-injection-d94f7c152681795c.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    For reporting bug fixes.
+    Use past tense for the problem and present tense for the fix and solution in the following format:
+    Format: <scope>: This fix resolves an issue where <ABC_bug> caused <XYZ_situation>.
```
Comment (P403n1x87) [inline releasenotes/notes/profiling-fix-protobuf-injection-d94f7c152681795c.yaml:6]:
```suggestion
```

**Review by P403n1x87 — APPROVED** [review]
LGTM!

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/settings/injection.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+class InjectionConfig(En):
+    __prefix__ = "dd.injection"
+
+    enabled = En.v(
+        bool,
+        "enabled",
+        default=False,
+        help_type="Boolean",
+        help="Whether the ddtrace library was injected into the application. This configuration will be applied "
+        "automatically if it is needed. Setting this manually not enable any features, it is for internal use "
```
[the remark this answers is not in the document]
Comment (sanchda) [reply ddtrace/settings/injection.py]: Good catch, tyvm

File: lib-injection/sitecustomize.py (hunk 2 above)
Comment (sanchda) [reply lib-injection/sitecustomize.py]: Makes total sense to me.  I made two related changes in 4ef8cbf:

* Replace the `DD_LIBRARY_INJECTION` with a direct setting of `DD_PROFILER_EXPORTER_LIBDD_ENABLE` (or whatever that crazy environment variable actually is)
* Remove the injection configuration and any references to it

I think we're good here, then!

**Review by Kyle-Verhoog — APPROVED** [review]
🚢 great stuff @sanchda!

**Comment by sanchda** [comment]
@Kyle-Verhoog and I had an offline discussion, where we agreed that `protobuf` actually needs to be stripped out of the inject Docker image and deb/rpm releases.  I made a best effort to do this, although I wanted to err on the side of caution and only remove the protobuf-related components rather than everything under `google/*`.  This leaves some empty directories right now.

```
/datadog-init $ find . -name '*protobuf*'
/datadog-init $ find . -name '*_upb*'
/datadog-init $ find . -name '*google*'
./ddtrace_pkgs/site-packages-ddtrace-py3.10-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.8-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.7-musllinux_1_1/google
./ddtrace_pkgs/site-packages-ddtrace-py3.9-musllinux_1_1/google
./ddtrace_pkgs/site-packages-ddtrace-py3.12-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.9-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.11-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.10-musllinux_1_1/google
./ddtrace_pkgs/site-packages-ddtrace-py3.7-manylinux2014/google
./ddtrace_pkgs/site-packages-ddtrace-py3.12-musllinux_1_1/google
./ddtrace_pkgs/site-packages-ddtrace-py3.11-musllinux_1_1/google
./ddtrace_pkgs/site-packages-ddtrace-py3.8-musllinux_1_1/google
```

**Pull request merged**